File: solutions/prospective/algorithms/CLRS/js/binary_search_rec.js
/**
 * Recursive Binary Search
 * 
 * Time Complexity: O(log n)
 * Space Complexity: O(log n) due to recursion stack
 * 
 * Returns the index of target in sorted array arr, or -1 if not found.
 */
// === MEMO START ===
function binarySearch(arr, target, low, high) {
    if (low > high) {
        return -1;
    }

    const mid = Math.floor((low + high) / 2);

    if (arr[mid] === target) {
        return mid;
    } else if (arr[mid] < target) {
        return binarySearch(arr, target, mid + 1, high);
    } else {
        return binarySearch(arr, target, low, mid - 1);
    }
}

File: solutions/prospective/algorithms/CLRS/c/binary_search_rec.c
/*
Recursive Binary Search

Time Complexity: O(log n)
Space Complexity: O(log n) due to recursion stack

Returns the index of target in sorted array arr, or -1 if not found.
*/
// === MEMO START ===
int binary_search(int *arr, int target, int low, int high) {
    if (low > high) {
        return -1;
    }

    int mid = (low + high) / 2;

    if (arr[mid] == target) {
        return mid;
    } else if (arr[mid] < target) {
        return binary_search(arr, target, mid + 1, high);
    } else {
        return binary_search(arr, target, low, mid - 1);
    }
}

